In [1]:
from pyspark.sql import SparkSession

In [2]:
#create a sparksession
spark=SparkSession.builder.appName('ass1').getOrCreate()

## read dataset and create one pyspark dataframe.

In [3]:
#now create a dataframe and iport data into dataframe
df=spark.read.csv(r"C:\Users\prem\Downloads\used_car_dataset_467142.csv", inferSchema=True, header=True)
df.show()

+--------------------+-------------------+-----------+----------+-----------+-------------------+
|            car_name|car_price_in_rupees| kms_driven| fuel_type|       city|year_of_manufacture|
+--------------------+-------------------+-----------+----------+-----------+-------------------+
|Hyundai Grand i10...|        ₹ 4.45 Lakh|  22,402 km|    Petrol|     Mumbai|               2016|
|Maruti Suzuki Alt...|        ₹ 2.93 Lakh|  10,344 km|    Petrol|    Kolkata|               2019|
|Tata Safari XZ Pl...|       ₹ 22.49 Lakh|  12,999 km|    Diesel|  Bangalore|               2021|
|Maruti Suzuki Cia...|        ₹ 6.95 Lakh|  45,000 km|    Petrol|      Thane|               2016|
|Jeep Compass Spor...|          ₹ 12 Lakh|  11,193 km|    Petrol|    Kolkata|               2019|
|Maruti Suzuki Eec...|        ₹ 3.72 Lakh|  42,906 km|    Petrol|    Chennai|               2017|
|Hyundai Creta SX ...|        ₹ 11.5 Lakh|  66,602 km|    Diesel|     Mumbai|               2017|
|Maruti Suzuki Ert..

i Fatched data Successfully but we can see this is too much noise value here i'll have to clean this one.

## Q3. Print the schema of the dataframe.

In [4]:
#let's check data type
df.printSchema()

root
 |-- car_name: string (nullable = true)
 |-- car_price_in_rupees: string (nullable = true)
 |-- kms_driven: string (nullable = true)
 |-- fuel_type: string (nullable = true)
 |-- city: string (nullable = true)
 |-- year_of_manufacture: integer (nullable = true)



every column data type has string because of noise value i'll have to remove these noise value after that i can convert data type

## Q4. find the shape of the dataframe.

In [10]:
#let's find the shape of this dataset
print(f"Shape:{df.count(), len(df.columns)}")

Shape:(2105, 6)


There are 2105 Rows and 6 Columns

## Q5. Print the 5 number summery of the dataframe.

In [12]:
#for 5 Number Summery 
df.describe().show()

+-------+--------------------+-------------------+-----------+----------+-----------+-------------------+
|summary|            car_name|car_price_in_rupees| kms_driven| fuel_type|       city|year_of_manufacture|
+-------+--------------------+-------------------+-----------+----------+-----------+-------------------+
|  count|                2105|               2105|       2105|      2105|       2105|               2105|
|   mean|                null|               null|       null|      null|       null| 2017.0228028503564|
| stddev|                null|               null|       null|      null|       null| 2.8180658246405534|
|    min|Audi A3 35 TDI Pr...|         ₹ 1.1 Lakh|1,00,000 km|       CNG|  Ahmedabad|               2004|
|    max|   Volvo XC90 D5 AWD|           ₹ 99,991|  99,941 km|Petrol + 1|Thiruvallur|               2022|
+-------+--------------------+-------------------+-----------+----------+-----------+-------------------+



1. Now here we can see every column count value is 2105 means there are non Null value
2. And kms_driven and fuel_type, car_price_in_rupees column data type is string that's why we can get every details;
3. but in the year_of_manufacture column we can notic that the min year was 2004 and latest year is 2022.

## Q6. Print the name of all columns. 

In [13]:
df.columns #this is the columns name list

['car_name',
 'car_price_in_rupees',
 'kms_driven',
 'fuel_type',
 'city',
 'year_of_manufacture']

## Q7. Print all the card having city = Mumbai and year_of_manufacture = 2016.

In [15]:
df.filter((df['city']=='Mumbai') & (df['year_of_manufacture']==2016)).show()

+--------------------+-------------------+----------+---------+------+-------------------+
|            car_name|car_price_in_rupees|kms_driven|fuel_type|  city|year_of_manufacture|
+--------------------+-------------------+----------+---------+------+-------------------+
|Hyundai Grand i10...|        ₹ 4.45 Lakh| 22,402 km|   Petrol|Mumbai|               2016|
|Volkswagen Vento ...|        ₹ 6.51 Lakh| 53,094 km|   Petrol|Mumbai|               2016|
|Ford Endeavour Ti...|       ₹ 24.75 Lakh| 83,000 km|   Diesel|Mumbai|               2016|
|       Honda City VX|        ₹ 7.45 Lakh| 42,218 km|   Petrol|Mumbai|               2016|
|Volkswagen Vento ...|        ₹ 7.25 Lakh| 45,689 km|   Petrol|Mumbai|               2016|
|Maruti Suzuki Cel...|        ₹ 4.45 Lakh| 24,595 km|   Petrol|Mumbai|               2016|
|Hyundai Creta 1.6...|        ₹ 7.45 Lakh| 62,461 km|   Petrol|Mumbai|               2016|
|Mercedes-Benz CLA...|       ₹ 19.25 Lakh| 34,000 km|   Petrol|Mumbai|               2016|

## Q8. Find all the cars of Maruti Suzuki.

In [18]:
df.filter(df['car_name'].contains('Maruti Suzuki')).show()

+--------------------+-------------------+----------+---------+---------+-------------------+
|            car_name|car_price_in_rupees|kms_driven|fuel_type|     city|year_of_manufacture|
+--------------------+-------------------+----------+---------+---------+-------------------+
|Maruti Suzuki Alt...|        ₹ 2.93 Lakh| 10,344 km|   Petrol|  Kolkata|               2019|
|Maruti Suzuki Cia...|        ₹ 6.95 Lakh| 45,000 km|   Petrol|    Thane|               2016|
|Maruti Suzuki Eec...|        ₹ 3.72 Lakh| 42,906 km|   Petrol|  Chennai|               2017|
|Maruti Suzuki Ert...|       ₹ 10.39 Lakh| 13,189 km|   Petrol|Bangalore|               2019|
|Maruti Suzuki Cia...|        ₹ 6.75 Lakh| 60,000 km|   Diesel|Ahmedabad|               2017|
|Maruti Suzuki Alt...|        ₹ 4.49 Lakh|  3,583 km|   Petrol|Bangalore|               2019|
|Maruti Suzuki Vit...|         ₹ 9.5 Lakh| 48,748 km|   Diesel|Hyderabad|               2019|
|Maruti Suzuki Cia...|         ₹ 9.9 Lakh|  7,755 km|   Petr

i used Contains Function.

## Q9. Crate new column which will contain age of the car.

In [23]:
#Now i'll have to create a function which will contain age of the car
df=df.withColumn('age', (2023-df['year_of_manufacture']))

In [24]:
df.show()

+--------------------+-------------------+-----------+----------+-----------+-------------------+---+
|            car_name|car_price_in_rupees| kms_driven| fuel_type|       city|year_of_manufacture|age|
+--------------------+-------------------+-----------+----------+-----------+-------------------+---+
|Hyundai Grand i10...|        ₹ 4.45 Lakh|  22,402 km|    Petrol|     Mumbai|               2016|  7|
|Maruti Suzuki Alt...|        ₹ 2.93 Lakh|  10,344 km|    Petrol|    Kolkata|               2019|  4|
|Tata Safari XZ Pl...|       ₹ 22.49 Lakh|  12,999 km|    Diesel|  Bangalore|               2021|  2|
|Maruti Suzuki Cia...|        ₹ 6.95 Lakh|  45,000 km|    Petrol|      Thane|               2016|  7|
|Jeep Compass Spor...|          ₹ 12 Lakh|  11,193 km|    Petrol|    Kolkata|               2019|  4|
|Maruti Suzuki Eec...|        ₹ 3.72 Lakh|  42,906 km|    Petrol|    Chennai|               2017|  6|
|Hyundai Creta SX ...|        ₹ 11.5 Lakh|  66,602 km|    Diesel|     Mumbai|     

## Q10. find the top 5 old car of Toyota company. Just print the car name and year.

In [29]:
df.filter(df['car_name'].contains('Toyota')).sort('age', ascending=False).select('car_name', 'year_of_manufacture').show(5)

+--------------------+-------------------+
|            car_name|year_of_manufacture|
+--------------------+-------------------+
|Toyota Corolla H2...|               2004|
|Toyota Corolla H2...|               2005|
|Toyota Innova 2.5...|               2009|
|Toyota Innova 2.5 G2|               2009|
|Toyota Corolla Al...|               2010|
+--------------------+-------------------+
only showing top 5 rows



These are the top 5 old car of toyota company.

## Q11. Find the cars those have Petrol as fuel_Type and city can be Noida or Delhi.

In [30]:
df.filter((df['fuel_Type'].contains('Petrol'))&(df['city'].isin('Noida', 'Delhi'))).show()

+--------------------+-------------------+----------+---------+-----+-------------------+---+
|            car_name|car_price_in_rupees|kms_driven|fuel_type| city|year_of_manufacture|age|
+--------------------+-------------------+----------+---------+-----+-------------------+---+
|Maruti Suzuki Bal...|        ₹ 6.39 Lakh| 34,559 km|   Petrol|Delhi|               2017|  6|
|  Tata Nexon XZ Plus|       ₹ 11.52 Lakh| 10,422 km|   Petrol|Delhi|               2021|  2|
| Hyundai Xcent S 1.2|        ₹ 3.95 Lakh| 62,568 km|   Petrol|Noida|               2014|  9|
|Maruti Suzuki S-P...|         ₹ 4.2 Lakh| 17,472 km|   Petrol|Delhi|               2020|  3|
|Hyundai Grand i10...|        ₹ 4.14 Lakh| 74,938 km|   Petrol|Delhi|               2016|  7|
| Honda City 1.5 S MT|        ₹ 2.85 Lakh| 82,000 km|   Petrol|Noida|               2009| 14|
|Maruti Suzuki Bal...|        ₹ 6.29 Lakh| 26,603 km|   Petrol|Delhi|               2019|  4|
|Maruti Suzuki Bal...|        ₹ 5.45 Lakh| 61,518 km|   Petr

## Q12. Find all the card those not are not belongs to fuel_Type of Diesel.

In [31]:
df.filter(df['fuel_Type'] != 'Diesel').show()

+--------------------+-------------------+-----------+----------+-----------+-------------------+---+
|            car_name|car_price_in_rupees| kms_driven| fuel_type|       city|year_of_manufacture|age|
+--------------------+-------------------+-----------+----------+-----------+-------------------+---+
|Hyundai Grand i10...|        ₹ 4.45 Lakh|  22,402 km|    Petrol|     Mumbai|               2016|  7|
|Maruti Suzuki Alt...|        ₹ 2.93 Lakh|  10,344 km|    Petrol|    Kolkata|               2019|  4|
|Maruti Suzuki Cia...|        ₹ 6.95 Lakh|  45,000 km|    Petrol|      Thane|               2016|  7|
|Jeep Compass Spor...|          ₹ 12 Lakh|  11,193 km|    Petrol|    Kolkata|               2019|  4|
|Maruti Suzuki Eec...|        ₹ 3.72 Lakh|  42,906 km|    Petrol|    Chennai|               2017|  6|
|Maruti Suzuki Ert...|       ₹ 10.39 Lakh|  13,189 km|    Petrol|  Bangalore|               2019|  4|
| Toyota Etios Liva G|        ₹ 3.75 Lakh|  67,815 km|    Petrol|     Mumbai|     

when i'm using this approch then i got some Diesel result also so i used anothe one approch.

In [32]:
df.filter(~(df['fuel_Type'].contains('Diesel'))).show()

+--------------------+-------------------+----------+---------+---------+-------------------+---+
|            car_name|car_price_in_rupees|kms_driven|fuel_type|     city|year_of_manufacture|age|
+--------------------+-------------------+----------+---------+---------+-------------------+---+
|Hyundai Grand i10...|        ₹ 4.45 Lakh| 22,402 km|   Petrol|   Mumbai|               2016|  7|
|Maruti Suzuki Alt...|        ₹ 2.93 Lakh| 10,344 km|   Petrol|  Kolkata|               2019|  4|
|Maruti Suzuki Cia...|        ₹ 6.95 Lakh| 45,000 km|   Petrol|    Thane|               2016|  7|
|Jeep Compass Spor...|          ₹ 12 Lakh| 11,193 km|   Petrol|  Kolkata|               2019|  4|
|Maruti Suzuki Eec...|        ₹ 3.72 Lakh| 42,906 km|   Petrol|  Chennai|               2017|  6|
|Maruti Suzuki Ert...|       ₹ 10.39 Lakh| 13,189 km|   Petrol|Bangalore|               2019|  4|
| Toyota Etios Liva G|        ₹ 3.75 Lakh| 67,815 km|   Petrol|   Mumbai|               2013| 10|
|Renault Kwid 1.0 ..